File: server/app/models/product.py
import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import Boolean, DateTime, ForeignKey, Index, Integer, Numeric, String, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import TimestampMixin, UUIDPrimaryKeyMixin, Base


class Product(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "products"

    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    platform: Mapped[str] = mapped_column(String(20), nullable=False)
    platform_product_id: Mapped[str] = mapped_column(String(255), nullable=False)
    product_name: Mapped[str] = mapped_column(String(500), nullable=False)
    shop_name: Mapped[str | None] = mapped_column(String(255))
    category: Mapped[str | None] = mapped_column(String(100))
    image_url: Mapped[str | None] = mapped_column(String(500))
    product_url: Mapped[str | None] = mapped_column(String(500))
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    last_collected_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    user: Mapped["User"] = relationship(back_populates="products")
    features: Mapped[list["ProductFeature"]] = relationship(back_populates="product", cascade="all, delete-orphan")

    __table_args__ = (
        UniqueConstraint("user_id", "platform", "platform_product_id", name="uq_user_platform_product"),
        Index("idx_products_user_id", "user_id"),
        Index("idx_products_platform", "platform"),
    )


class ProductFeature(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "product_features"

    product_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("products.id", ondelete="CASCADE"), nullable=False)
    price: Mapped[Decimal | None] = mapped_column(Numeric(12, 2))
    original_price: Mapped[Decimal | None] = mapped_column(Numeric(12, 2))
    sales_count: Mapped[int | None] = mapped_column(Integer)
    monthly_sales: Mapped[int | None] = mapped_column(Integer)
    rating: Mapped[Decimal | None] = mapped_column(Numeric(3, 2))
    review_count: Mapped[int | None] = mapped_column(Integer)
    favorite_count: Mapped[int | None] = mapped_column(Integer)
    stock_status: Mapped[str | None] = mapped_column(String(20))
    extra_features: Mapped[dict] = mapped_column(JSONB, default={}, server_default="{}")
    collected_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    source: Mapped[str] = mapped_column(String(20), nullable=False, default="local")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    product: Mapped["Product"] = relationship(back_populates="features")

    __table_args__ = (
        Index("idx_product_features_product_id", "product_id"),
        Index("idx_product_features_collected_at", "collected_at"),
    )

File: client/src/renderer/stores/product.ts
import { defineStore } from "pinia";
import { ref, computed } from "vue";

export interface Product {
  id: string;
  platform: string;
  platform_product_id: string;
  product_name: string;
  shop_name: string | null;
  category: string | null;
  image_url: string | null;
  product_url: string | null;
  last_collected_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ProductFeature {
  id: string;
  product_id: string;
  price: number | null;
  original_price: number | null;
  sales_count: number | null;
  monthly_sales: number | null;
  rating: number | null;
  review_count: number | null;
  favorite_count: number | null;
  stock_status: string | null;
  collected_at: string;
  source: string;
}

export const useProductStore = defineStore("product", () => {
  const products = ref<Product[]>([]);
  const currentProduct = ref<Product | null>(null);
  const features = ref<ProductFeature[]>([]);
  const loading = ref(false);
  const error = ref<string | null>(null);
  const filter = ref<{ platform?: string; search?: string }>({});

  const productCount = computed(() => products.value.length);
  const activeProducts = computed(() => products.value.filter((p) => p.last_collected_at));

  async function fetchProducts() {
    loading.value = true;
    error.value = null;
    try {
      const result = await window.electronAPI.invoke("datamart:list-products", filter.value.platform, 100);
      products.value = result as Product[];
    } catch (err) {
      error.value = String(err);
    } finally {
      loading.value = false;
    }
  }

  async function fetchProductDetail(productId: string) {
    loading.value = true;
    try {
      const result = await window.electronAPI.invoke("datamart:get-product", productId);
      currentProduct.value = result as Product;
    } catch (err) {
      error.value = String(err);
    } finally {
      loading.value = false;
    }
  }

  async function fetchFeatures(productId: string) {
    try {
      const result = await window.electronAPI.invoke("storage:query", "SELECT * FROM product_features WHERE product_id = ? ORDER BY collected_at DESC LIMIT 30", [productId]);
      features.value = (result as ProductFeature[]).map((f: ProductFeature) => ({
        ...f,
        extra_features: typeof (f as Record<string, unknown>).extra_features === "string"
          ? JSON.parse((f as Record<string, unknown>).extra_features as string)
          : (f as Record<string, unknown>).extra_features,
      }));
    } catch (err) {
      error.value = String(err);
    }
  }

  function setFilter(newFilter: { platform?: string; search?: string }) {
    filter.value = { ...filter.value, ...newFilter };
  }

  return {
    products,
    currentProduct,
    features,
    loading,
    error,
    filter,
    productCount,
    activeProducts,
    fetchProducts,
    fetchProductDetail,
    fetchFeatures,
    setFilter,
  };
});
